core/src/odbc_uri.rs

```diff
@@ -37,6 +37,7 @@ lazy_static! {
     .case_insensitive(true)
     .build()
     .unwrap();
+    static ref AUTH_SRC_REGEX: Regex = Regex::new(r"authSource=([a-zA-Z_0-9]*)").unwrap();
```

Comment:
You want AuthSource to be case insensitive. Maybe use the RegexBuilder because Regex::new seems to create a case-sensitive regex., or create a group for `authSrc` and set the case-sensitivity there.
And for the value, I don't think that you can limit the authSource to alphanumeric characters. One of the reason is that in the description of the authSource option it says `The PLAIN (LDAP), GSSAPI (Kerberos), and MONGODB-AWS (IAM) authentication mechanisms require that [authSource](https://www.mongodb.com/docs/manual/reference/connection-string/#mongodb-urioption-urioption.authSource) be set to $external`.
 I also did not see anywhere that it had to be restricted to alphanumeric characters.  It does appear so from [this page](https://www.mongodb.com/docs/manual/reference/limits/).

[nit] Another thing that I can suggest is to name your capture group. This way, no need to worry about the capture index, does it start at 0 or 1, etc...

In the end, I would use this:
```
lazy_static! {
    static ref AUTH_SRC_REGEX: Regex = RegexBuilder::new(r#"[&?]authSource=(?P<authSrc>[^&]*)"#)
        .case_insensitive(true)
        .build()
        .unwrap();
}
and grab the value with
```
let auth_src_val = AUTH_SRC_REGEX
            .captures(&uri)
            .and_then(|cap| cap.name("authSrc").map(|authSrc| authSrc.as_str()));
```
It support all characters except '&' because this one marks the next option (if the authSource name contains a '&' it will have to be URL encoded). I think that it is OK to let the Rust driver or MDB server handle the validity of the value.
This way, it purposefully allows '$' and will support `$external` (and exotic source names too)
```
My unit tests seem Ok.
Here is the list I tested out:
```
       "mongodb://localhost/?authSource=authDB",
        "mongodb://localhost/",
        "mongodb://localhost/?auTHSource=aut:hD@B",
        "mongodb://localhost/?auTHSource=aut:hD@B&anotherOpt=tgg#fed",
        "mongodb://uid:pwd@localhost/?authSource=$external&anotherOpt=tgg#fed",
        "mongodb://localhost/?something=test&auTHSource=aut:hD@B",
        "mongodb://localhost/?something=test&tlsCertificateKeyFilePassword='myauthSource=aut:hD@B'&authSource=jfhbgvhj"
```

Reply:
Yes, this is correct. & should be percent encoded if used!